test/kt/single_pass_copy_if.cpp

```diff
@@ -0,0 +1,261 @@
+// -*- C++ -*-
+//===-- single_pass_copy_if.cpp -------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../support/test_config.h"
+
+#include <oneapi/dpl/experimental/kernel_templates>
+
+#if LOG_TEST_INFO
+#    include <iostream>
+#endif
+
+#if _ENABLE_RANGES_TESTING
+#    include <oneapi/dpl/ranges>
+#endif
+
+#include "../support/utils.h"
+#include "../support/sycl_alloc_utils.h"
+#include "../support/scan_serial_impl.h"
+
+#include "esimd_radix_sort_utils.h"
+
+#include <random>
+#include <algorithm>
+#include <numeric>
+#include <vector>
+#include <cstdint>
+#include <type_traits>
+
+inline const std::vector<std::size_t> copy_if_sizes = {
+    1,       6,         16,      43,        256,           316,           2048,
+    5072,    8192,      14001,   1 << 14,   (1 << 14) + 1, 50000,         67543,
+    100'000, 1 << 17,   179'581, 250'000,   1 << 18,       (1 << 18) + 1, 500'000,
+    888'235, 1'000'000, 1 << 20, 10'000'000};
+
+template <typename T>
+struct __less_than_val
+{
+    const T __val;
+    bool
+    operator()(const T& __v) const
+    {
+        return __v < __val;
+    }
+};
+
+template <typename T>
+auto
+generate_copy_if_data(T* input, std::size_t size, std::uint32_t seed)
+{
+    // Integer numbers are generated even for floating point types in order to avoid rounding errors,
+    // and simplify the final check
+    using substitute_t = std::conditional_t<std::is_signed_v<T>, std::int64_t, std::uint64_t>;
+
+    std::default_random_engine gen{seed};
+    substitute_t start = std::is_signed_v<T> ? -10 : 0;
+    substitute_t end = std::is_signed_v<T> ? 10 : 20;
+    std::uniform_int_distribution<substitute_t> dist(start, end);
+    std::generate(input, input + size, [&] { return dist(gen); });
+}
+
+#if _ENABLE_RANGES_TESTING
+template <typename T, typename Predicate, typename KernelParam>
+void
+test_all_view(sycl::queue q, std::size_t size, Predicate pred, KernelParam param)
+{
+#    if LOG_TEST_INFO
+    std::cout << "\ttest_all_view(" << size << ") : " << TypeInfo().name<T>() << std::endl;
+#    endif
+    std::vector<T> input(size);
+    generate_copy_if_data(input.data(), size, 42);
+    std::vector<T> ref(input);
+    std::vector<T> out_ref(size);
+    sycl::buffer<T> buf_out(input.size());
+    std::size_t num_copied = 0;
+    sycl::buffer<std::size_t> buf_num_copied(&num_copied, 1);
+    auto out_end = std::copy_if(std::begin(ref), std::end(ref), std::begin(out_ref), pred);
+    std::size_t num_copied_ref = out_end - std::begin(out_ref);
+    {
```

Comment:
Let's get rid of this scope. The check below does not need `buf` to be copied to the host.